sw/device/sca/lib/aes.c

```diff
@@ -0,0 +1,224 @@
+  state[1] ^= temp[1] ^ temp[2];
+  state[2] ^= temp[2] ^ temp[3];
+  state[3] ^= temp[3] ^ temp[0];
+
+  return;
+}
+
+void aes_transpose_to_32(uint8_t *in_data, uint32_t *out_data) {
+  out_data[0] = (uint32_t)in_data[0] | ((uint32_t)in_data[4] << 8) |
+                ((uint32_t)in_data[8] << 16) | ((uint32_t)in_data[12] << 24);
+  out_data[1] = (uint32_t)in_data[1] | ((uint32_t)in_data[5] << 8) |
+                ((uint32_t)in_data[9] << 16) | ((uint32_t)in_data[13] << 24);
+  out_data[2] = (uint32_t)in_data[2] | ((uint32_t)in_data[6] << 8) |
+                ((uint32_t)in_data[10] << 16) | ((uint32_t)in_data[14] << 24);
+  out_data[3] = (uint32_t)in_data[3] | ((uint32_t)in_data[7] << 8) |
+                ((uint32_t)in_data[11] << 16) | ((uint32_t)in_data[15] << 24);
+}
+
+void aes_transpose_from_32(uint32_t *in_data, uint8_t *out_data) {
+  out_data[0] = (uint8_t)(in_data[0] & 0xff);
+  out_data[1] = (uint8_t)(in_data[1] & 0xff);
+  out_data[2] = (uint8_t)(in_data[2] & 0xff);
+  out_data[3] = (uint8_t)(in_data[3] & 0xff);
+  out_data[4] = (uint8_t)(in_data[0] >> 8) & 0xff;
+  out_data[5] = (uint8_t)(in_data[1] >> 8) & 0xff;
+  out_data[6] = (uint8_t)(in_data[2] >> 8) & 0xff;
+  out_data[7] = (uint8_t)(in_data[3] >> 8) & 0xff;
+  out_data[8] = (uint8_t)(in_data[0] >> 16) & 0xff;
+  out_data[9] = (uint8_t)(in_data[1] >> 16) & 0xff;
+  out_data[10] = (uint8_t)(in_data[2] >> 16) & 0xff;
+  out_data[11] = (uint8_t)(in_data[3] >> 16) & 0xff;
+  out_data[12] = (uint8_t)(in_data[0] >> 24) & 0xff;
+  out_data[13] = (uint8_t)(in_data[1] >> 24) & 0xff;
+  out_data[14] = (uint8_t)(in_data[2] >> 24) & 0xff;
+  out_data[15] = (uint8_t)(in_data[3] >> 24) & 0xff;
+}
+
+void aes_key_expand(uint8_t *round_key, uint8_t *rcon) {
+  uint8_t temp[4];
+  uint8_t old_key[16];
+
+  // copy key to temp
+  memcpy(old_key, round_key, 16);
+
+  // shift last word
+  temp[0] = old_key[13];
+  temp[1] = old_key[14];
+  temp[2] = old_key[15];
+  temp[3] = old_key[12];
+
+  // sub bytes in last word
+  temp[0] = sbox[temp[0]];
+  temp[1] = sbox[temp[1]];
+  temp[2] = sbox[temp[2]];
+  temp[3] = sbox[temp[3]];
+
+  // update rcon
+  aes_rcon_next(rcon);
+
+  // get new words
+  round_key[0] = temp[0] ^ old_key[0] ^ *rcon;
+  round_key[1] = temp[1] ^ old_key[1];
+  round_key[2] = temp[2] ^ old_key[2];
+  round_key[3] = temp[3] ^ old_key[3];
+
+  for (size_t i = 0; i < 12; i++) {
+    round_key[i + 4] = round_key[i] ^ old_key[i + 4];
+  }
+
+  return;
+}
+
+void aes_rcon_next(uint8_t *rcon) {
+  // rcon cannot be 0
+  if (*rcon) {
+    // update round constant
+    *rcon = mul2[*rcon];
+  } else {
+    // init round constant to first-round value
+    *rcon = 0x1;
+  }
+
+  return;
+}
+
+void aes_key_schedule(uint32_t *round_keys, uint8_t *key) {
+  // Derives all round keys for AES128
+  // Each key is storred in 4 32-bit words in a transposed-state form.
+  size_t num_rounds = 10;
+  uint8_t rcon = 0;
+  uint8_t key_temp[16];
+  uint32_t key_temp_32[4];
```

Comment:
Here we should use constants/enums for the literals as well.

Reply:
Agreed. I've changed it now.